### PR TITLE
Large updates to wiki-refresh for 2019 dataset

### PR DESCRIPTION
**Pull request by frankduncan**
@kfogel this is just one large commit, but I can attempt to break it apart if you like.  Outside of that, we can do the normal review process, or just merge it into master and move forward with the project.

The updates in this include:

* Move fix-csv to compose-csvs
* Move csv munging code out of wiki-refresh base script into python, now
  located in compose-csvs
* Call csv2wiki via python3 module loading
* Update macfound-internal-csv2wiki-config.tmpl to include judge data

More generally, this update takes three different csvs and puts them
together into one set of data that's culled by which proposals have been
reviewed favorably, then adds judge review data to those proposals to
output to the wiki instance.  This work is all done in one place, with
wiki-refresh now being a small combiner script with variables about data
placement.

### CODE REVIEW

**Comment by kfogel** [comment]
Happy to review as one large commit; doing so now.  I will be interrupted by a couple of meetings soon, so the review will take a few hours of clock time.

**Comment by kfogel** [comment]
Oh, ouch, you rebased while I was reviewing :-).

File: compose-csvs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+                    "sum_of_scores_normalized": row[12],
+                    "judge_data": "",
+                    }
+        judge_data[application_id]["judge_data"] += \
+                row[14] + "(" + row[15] + ", norm: " + row[16] + "): " + row[17] + "\n\n"
+
+    return judge_data
+
+def compose_csvs(proposals, admin_review, judge_eval, pare=None):
+    """Write a composed version of PROPSALS (a CSV file), ADMIN_REVIEW
```
Comment (kfogel) [inline compose-csvs]: typo: "PROPSALS"

File: compose-csvs [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    """Print the ROWS to stdout"""
+
+    csv_writer = csv.writer(sys.stdout,
+                            delimiter=',', quotechar='"', lineterminator="\n")
+
+    csv_writer.writerow(header_row)
+    for row in rows:
+        csv_writer.writerow(row)
+
+def process_judge_data(csv_reader):
```
[the remark this answers is not in the document]
Comment (kfogel) [reply compose-csvs]: Later I saw code that makes me think this returns a dictionary keyed by Review Number (a.k.a. Application Number).  That should be documented.

File: wiki-refresh [hunk 3]
```diff
@@ -55,44 +46,15 @@ if [ "${DATA_DIR}" = "" ]; then
...
-   git clone https://github.com/OpenTechStrategies/csv2wiki.git)
-  CSV2WIKI="${SOURCES_ROOT}/csv2wiki/csv2wiki"
-fi
-if [ ! -x ${CSV2WIKI} ]; then
-  echo "ERROR: Unable to find or clone 'csv2wiki'."
-  exit 1
-fi
-# Test that csv2wiki actually works.
-if ${CSV2WIKI} --help > /dev/null; then
+python3 -m csv2wiki --version >/dev/null 2>&1 || { echo "csv2wiki not installed, make sure 'python3 -m csv2wiki' works required, aborting" ; exit 1; }
```
Comment (kfogel) [inline wiki-refresh]: Does this code now assume that it's running from the top of the `csv2wiki` tree?

File: compose-csvs [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+#   open source software, so please modify it to suit your needs.        #
+#                                                                        #
+##########################################################################
+
+__doc__ = """\
+Compose all of the MacArthur Foundation 2019 Proposal CSV files.
+
+Usage:
+
+  $ compose-csvs PROPOSALS_CSV ADMIN_REVIEW_CSV JUDGE_EVALUATION_CSV
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply compose-csvs]: On the todo list (should be easy fix next week).

File: compose-csvs [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+import warnings
+import string
+from bs4 import BeautifulSoup
+
+# Convert "<foo>&nbsp;<bar>" and "</foo>&nbsp;<bar>" to
+# to "<foo> <bar>" and "</foo> <bar>" respectively.  This is
+# because those particular instances of "&nbsp;" in the data are
+# not very convincing, and they create noise when we're looking
+# for unnecessary escaping elsewhere.
+intertag_nbsp_re = re.compile('(?m)(</?[a-z]+>)&nbsp;(<[a-z]+>)')
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply compose-csvs]: So this regex used to be outside of the loop which then used it in a long method.  I separated the method out to fix_cell, but felt like recompiling it each time that method was called was bad?  Is there some pythonic way to do what you're talking about here?  In CL I would do:

```
(let
 ((re (compile-some-regex)))
 (defun fix-cell (cell) ...))
```

File: compose-csvs [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+
+bullets_re = re.compile("^•", re.MULTILINE)
+"""Run fixes on the data in CELL"""
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply compose-csvs]: See above for best way to combine these pieces of code?

File: compose-csvs [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    # bold and return.
+    if re.sub(r"\W", "", soup.get_text()) == "":
+        return re.sub("</?strong>", "", html)
+
+    # OTOH, if the non-bold stuff contained letters or numbers, maybe
+    # there's real content there, which means the html was a mix of
+    # bold and non-bold text.  Better to leave it alone.
+    return html
+
+def form_well(html):
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply compose-csvs]: Want me to add some status output if things did get cleaned as part of the running program?

File: compose-csvs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kfogel) [reply compose-csvs]: I was mainly talking about the comment placement -- the comment talks about what something *does*, and things that do something are methods and functions, so the comment probably should be near one of those.  But the regexp could be moved nearer its caller too.

I will do so in followup commit.

File: compose-csvs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kfogel) [reply compose-csvs]: Will do in a followup commit.

File: compose-csvs (hunk 7 above)
[the remark this answers is not in the document]
Comment (kfogel) [reply compose-csvs]: Yeah; it doesn't block this merge, but we should have a diagnostic mode so we can see how much of this cleanup still applies to the new data.

**Pull request merged**